### ISSUE TITLE
Support relative imports in quick-fixes + auto-imports

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#32916]
I'm working in `flutter_tools` and the imports inside `lib/src` are all relative paths. However, the quick-fixes keep wanting to add imports as `package:` (I believe this is unwanted, and `package:` imports are recommended only from outside the package).

<img width="801" alt="screen shot 2018-04-18 at 10 25 51" src="https://user-images.githubusercontent.com/1078012/38923622-f4a5e9ca-42f2-11e8-885f-e666086a37bc.png">

In this case the desired import is `import 'emulator.dart';`.

**Comment by bwilkerson** [comment]
Using both relative and `package:` URIs to reference libraries in the defining package can lead to problems.

There are two valid approaches to prevent those problems, and analyzer has always taken the approach of always using `package:` URIs everywhere. So, in that sense at least, it's working as intended. I realize that some people prefer relative paths because they are often (though not always) shorter, but that's not the default that was chosen. (Using a `package:` URI has other advantages, such as making it impossible to accidentally reference libraries outside of the `lib` directory from within the `lib` directory.)

What we probably ought to do is look for the existing pattern in a package and figure out which to use. Or possibly add an option to the analysis options file. But both require more work than we have time to take on right now.

**Comment by DanTup** [comment]
The weird thing is, sometimes the analyzer does suggest relative paths and sometimes it doesn't (I haven't paid enough attention to know if there's a pattern).

This isn't my codebase so I don't know why it's this way here (`flutter_tools`) but I do kinda dislike `package:` imports (not only because they're usually longer - seems you're more likely to import things that are "close" to you, but they also break less if you move things around - eg. if files import their siblings and you rename the folder they're in, their imports of each other don't break).

It's not a huge deal, it's just really easy to import stuff "the wrong way" and break things. Ideally it wouldn't matter (I think there was some discussion somewhere about making this so, but I don't know the scope of it, of whether it's actually likely to happen).

**Comment by bwilkerson** [comment]
> ... they also break less if you move things around ...

On the other hand, if you move individual files, then the `package:` imports in the file being moved don't break (the imports of the file being moved will break either way). I suppose it depends on the style of refactoring you're more likely to use.

**Comment by escamoteur** [comment]
now having the problem again. it only offers me relative path and no package import

**Comment by escamoteur** [comment]
it seems to be problem if the two dependent files are in parallel folders

like

/lib/service/service.dart
/lib/hompage/homepage.dart

in homepage.dart it only offers me the relative path to service.dart

**Comment by DanTup** [comment]
@bwilkerson 

> Can you tell me what you were doing at that point (refactoring, quick fix
for some error, etc)?

I created the class `EmulatorManager` today and was adding it to `packages/flutter_tools/lib/src/context_runner.dart` in the flutter repo. I added the line you can see with the errors, and then used quick-fix to add the import for it.

**Comment by DanTup** [comment]
I fell foul of this again in the PR above.. I used the quick-fix in VS Code and inserted imports using `package:` which is different to all the others in DevTools. There are also a number of other issues closed as dupes of this linked above.

On the website it says to prefer relative paths for within your package:

https://dart.dev/guides/language/effective-dart/usage#prefer-relative-paths-when-importing-libraries-within-your-own-packages-lib-directory

though since the server is often doing the opposite, I'm not sure which is most common. @pq - I'm told you may have a tool that can scan a lot of OSS code that may be able to get numbers on this? If relative paths is the most common and the way suggested on the site, there's probably value in having the server suggest that first?

**Comment by Zorgatone** [comment]
Still nothing? This is much needed :(

**Comment by bwilkerson** [comment]
Just to update, I believe that the quick fixes should now come in both varieties, allowing users to select the form they prefer. The auto-import on completion support has not been started.

**Comment by DanTup** [comment]
@bwilkerson for the auto-import work, will it result in the items being duplicated, or will there be some way for the server to discover the users preference? (I think there are already some lints around this topic, so perhaps they could be used to indicate a preference?). If so, it'd be nice if the quick-fix that matched the preference was also above the other in the list (or even, only that one shown?).

**Comment by bwilkerson** [comment]
We're currently showing both quick fixes. I don't know of any use case for users wanting to be able to choose between them (you basically need to always use one or always use the other to avoid importing the same library but have Dart think it's two identical but distinct libraries), but, as you noted, we need is a way to discover the user's preference. The enablement of lints is one possibility, but only works if users have enabled one lint or the other. Another possibility we've discussed is to look at existing imports to see which style is used.

I think we'll need something like that in order to reasonably improve code completion. We don't want to have two completion suggestions for every importable identifier.

**Comment by bsutton** [comment]
I'm using effective dart which contains the lint prefer_relative_imports.
You can see the merge for this rule here:
tenhobi/effective_dart#21

It would be useful if you could detect this lint and act accordingly.

I'm also not always being offered a relative path.
Doing some experimentation it looks like if the import is in or below the same folder then the 'relative import' is offered.
If the path is above the current folder and the relative path would contain '..' then only the package import is offered.

I would suggest:
If the source file is under 'lib' then any file also under lib should be offered as relative.
If the source file is under 'example' then any file under example should be offered as relative.
Same for test.
I'm not certain about bin. I'm guessing that you are not meant to have a folder structure under the bin directory just a set of libraries each with their own 'main'.

**Comment by danielgomezrico** [comment]
Is there any work on behalf of this? I think that it is also related to this https://github.com/dart-lang/sdk/issues/44626 which already works on Kotlin world, it's a developer holy gift to be able to have better import hints from the IDE 🙏🏻

**Comment by PiN73** [comment]
Is there any way to make auto-imports relative by default in Android Studio? E.g. when I type MyHomeP and press Tab

**Comment by YuriHeiko** [comment]
Any updates on this issue?

**Comment by srawlins** [comment]
It should be fixed in Dart 2.14.

**Comment by PiN73** [comment]
Auto-imports still use absolute paths both in Android Studio and VS Code. Is it analysis server or IDE plugin's issue? 

## Reproduction steps

analysis_options.yaml
```yaml
linter:
  rules:
    prefer_relative_imports: true
```

hello.dart
```dart
const hello = 'hello';
```

main.dart
```dart
void main() {
  
}
```

Type `hell` inside main function and press <kbd>Tab</kbd>

## Expected results

After <kbd>Tab</kbd> is pressed, this line should appear at the top of main.dart
```dart
import 'hello.dart';
```

## Actual results

After <kbd>Tab</kbd> is pressed, this line appears at the top of main.dart
```dart
import 'package:myapp/hello.dart';
```

## Versions

```
% flutter --version
Flutter 2.3.0-24.0.pre • channel dev • https://github.com/flutter/flutter.git
Framework • revision 2b9537c783 (11 days ago) • 2021-06-11 22:11:53 -0700
Engine • revision ddbac024d1
Tools • Dart 2.14.0 (build 2.14.0-188.0.dev)
```

macOS 11.3.1 20E241 darwin-x64

Android Studio 4.2.1
Dart plugin 202.8488
Flutter plugin 57.0.2

Visual Studio Code 1.56.2
Dart extension v3.23.1
Flutter extension v3.23.0

**Comment by DanTup** [comment]
@srawlins I sneakily included two things in the title of this issue (quick-fixes + auto-imports), but I think the fix above may have only applied to the first?

**Comment by bwilkerson** [comment]
Yes, the auto-import via code completion is still not addressed. That's blocked on deciding how to let users specify which style they prefer.

**Comment by DanTup** [comment]
The change above uses the `prefer_relative_imports` lint - could that be the same here (assume `package:` URIs if it's not enabled - which is the current behaviour - and use relative paths if it's enabled)?

**Comment by bwilkerson** [comment]
I'm sorry that we're using `prefer_relative_imports` for quick fixes. I think it's a huge mistake to _require_ users to enable a lint in order to express the behavior they'd like to see for code completion and quick fixes. (I think it's fine to use it as a signal, just not the _only_ signal.)

But I suppose that _some_ signal is better than _no_ signal, so I suppose that we ought to use the lint as a signal for code completion now that we're already using as such for quick fixes.

**Comment by lrhn** [comment]
Another hint could be if the file already contains relative or absolute imports, then be consistent with that.
(Beats having to fix the new imports manually).

**Comment by tjx666** [comment]
Why so basic features still not provide?

**Comment by bwilkerson** [comment]
> Why so basic features still not provide?

Because we have limited resources and other tasks were given a higher priority.

> Every time I need to modify `package import` to `relative import` manually after using auto import ...

I empathize with your frustration. In case it helps, we do have an assist to convert from absolute (package) to relative and another to convert in the opposite direction. That might make it a tiny bit easier.

**Comment by jangruenwaldt** [comment]
Please fix hahaha this is quite painful 😢

**Comment by Zeswen** [comment]
+1

**Comment by VovaMedvediev** [comment]
+

**Comment by VovaMedvediev** [comment]
+++

**Comment by bwilkerson** [comment]
As far as I'm aware, both issues have now been addressed, so I'm going to close this. We can re-open it, or open a new issue, if there are additional pieces of functionality that aren't working as desired.

**Comment by danielgomezrico** [comment]
@bwilkerson any chance to take a look to this one too? https://github.com/dart-lang/sdk/issues/44626

**Comment by bwilkerson** [comment]
Yes, we definitely need to fix that issue as well. But I believe that the changes @DanTup made will also apply in that case. That is, when we are able to produce a quick fix for references to extension methods, the fix will correctly support relative imports.

**Comment by bsutton** [comment]
@bwilkerson I don't believe this has been fixed.

I'm running dart 2.15 in dart-code v3.29.0.

If I use a class that is missing an import and open the quick fix then I am not always offered the relative path option.

I think this relates to where the import is relative to my current library.

For this import I was offered both
import '../batman_settings.dart';

For this import I was only offered the package import
import '../hive/hive_store.dart';

I have the prefer_relative_imports lint enabled.

FYI: @DanTup 

This is a really frustrating little bug.

**Comment by DanTup** [comment]
@bsutton what was the location of the file you were adding the import to? I tried a few places but seem to get both (using Dart 2.15.0):

<img width="413" alt="Screenshot 2021-12-10 at 11 10 04" src="https://user-images.githubusercontent.com/1078012/145565074-67e690e2-4a32-490a-a63c-44c3ab4b11ce.png">

If you can help me repro, I'll take a look. Thanks!

**Comment by bsutton** [comment]
@DanTup  the project is on github:
https://github.com/noojee/batman

The file was lib/src/commands/baseline.dart

1) clone the  project
2) run pub get
3) remove hive_store import from baseline.dart 

Try to re-add hive_store with a quick fix.

FWIW: the problem occurs on windows and linux.

**Comment by DanTup** [comment]
@bsutton thanks! I thought it seemed familiar - I hit the same while working on the change above (3dbd06bdfce438e76d4b4022d550171cb85a8ee3). There was an issue in [the function to compute the relative path](https://github.com/dart-lang/sdk/commit/3dbd06bdfce438e76d4b4022d550171cb85a8ee3#diff-1c81936e2696eef0abad669d4091ee6ddcba18e0570907656c7f4ac8e29bd3c6L153) which required one folder to be inside the other (which fails for `batman/lib/src/hive` vs `batman/lib/src/commands`). I suspect it was intended to check both files were inside the `lib` folder, which is essentially what [the replacement](https://github.com/dart-lang/sdk/commit/3dbd06bdfce438e76d4b4022d550171cb85a8ee3#diff-bcbaac9f974f8f96ae8bba74035a5782f1063179c633dc93cf8c5876919e21e0R10) does.

Unfortunately that change did not make the v2.15.0 branch so won't show up until a future SDK release (it's likely already available in the dev channel etc.).

**Comment by bsutton** [comment]
Good news then.

As always thanks for being so responsive.
